UM/Resources.py

```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
+        if platform.system() == "Windows":
+            # it used to be in LOCALAPPDATA on Windows
+            config_root_list.append(os.getenv("LOCALAPPDATA"))
+
+        config_root_list = [os.path.join(n, cls.ApplicationIdentifier) for n in config_root_list]
+        return config_root_list
+
+    @classmethod
+    def getDataStorageRootPath(cls):
+        # Returns the path where we store different versions of app data
+        data_path = None
+        if platform.system() == "Linux":
             try:
-                xdg_data_home = os.environ["XDG_DATA_HOME"]
+                data_path = os.environ["XDG_DATA_HOME"]
             except KeyError:
-                xdg_data_home = os.path.expanduser("~/.local/share")
-            cls.__data_storage_path = os.path.join(xdg_data_home, storage_dir_name)
+                data_path = os.path.expanduser("~/.local/share")
+        return data_path
 
-            xdg_cache_home = ""
+    @classmethod
+    def getCacheStorageRootPath(cls):
+        # Returns the path where we store different versions of app configurations
+        cache_path = None
+        if platform.system() == "Windows":
+            cache_path = os.getenv("LOCALAPPDATA")
+        elif platform.system() == "Darwin":
+            cache_path = None
+        elif platform.system() == "Linux":
             try:
-                xdg_cache_home = os.environ["XDG_CACHE_HOME"]
+                cache_path = os.environ["XDG_CACHE_HOME"]
             except KeyError:
-                xdg_cache_home = os.path.expanduser("~/.cache")
-            cls.__cache_storage_path = os.path.join(xdg_cache_home, storage_dir_name)
-        else:
-            cls.__config_storage_path = "."
+                cache_path = os.path.expanduser("~/.cache")
 
-        if not cls.__data_storage_path:
-            cls.__data_storage_path = cls.__config_storage_path
+        return cache_path
 
-        if not cls.__cache_storage_path:
+    @classmethod
+    def __initializeStoragePaths(cls):
+        # use nested structure: <app-name>/<version>/...
+        if cls.ApplicationVersion == "master" or cls.ApplicationVersion == "unknown":
+            storage_dir_name = os.path.join(cls.ApplicationIdentifier, cls.ApplicationVersion)
+        else:
+            from UM.Version import Version
+            version = Version(cls.ApplicationVersion)
+            storage_dir_name = os.path.join(cls.ApplicationIdentifier, "%s.%s" % (version.getMajor(), version.getMinor()))
+
+        # config is saved in "<CONFIG_ROOT>/<storage_dir_name>"
+        cls.__config_storage_path = os.path.join(Resources.getConfigStorageRootPath(), storage_dir_name)
+
+        # data is saved in
+        #  - on Linux: "<DATA_ROOT>/<storage_dir_name>"
+        #  - on other: "<CONFIG_DIR>" (in the config directory)
+        data_root_path = Resources.getDataStorageRootPath()
+        cls.__data_storage_path = cls.__config_storage_path if data_root_path is None else \
```

Comment:
Please just write this as

```
if data_root_path is None:
  cls.__data_storage_path = cls.__config_storage_path
else:
  cls.__data_storage_path = os.path.join(data_root_path, storage_dir_name)
```

This prevents ugly escaping of newlines and is generally much more readable.